papyri/take2.py

```diff
@@ -159,14 +159,33 @@ class Leaf(Node):
 
 @register(4027)
 class SubstitutionDef(Node):
+    """
+    A Substitution Definition should be in the form of
+
+    .. raw:: rst
+
+        .. |value| inline_directive:: text to be inserted
+
+    Currently, the inline_directive can only be ``replace`` or ``image``. In the
+    future, we want to support any inline directives, including custom
+    user-defined directives.
+    """
+
     value: str
-    children: List[Union[MMystDirective, UnprocessedDirective]]
+    children: List[Union[ReplaceNode, MImage]]
 
     def __init__(self, value, children):
         self.value = value
         assert isinstance(children, list)
-        self.children = children
-        pass
+        assert len(children) == 1
+        assert isinstance(children[0], UnprocessedDirective)
+
+        if children[0].name == "image":
+            self.children = [MImage(url=children[0].args, alt="")]
+        elif children[0].name == "replace":
+            self.children = [ReplaceNode(value=self.value, text=children[0].args)]
+        else:
```

Comment:
Here is a quick search of what kind of subst we can find:

```
$ rg '^\.\. \|.+\| ' | cut -f3 -d'|'  | cut -f 1 -d':' |tr -d ' '| sort | uniq -c | sort -n
htb/2023-sept-responder/Responder/certs/responder.key: Permission denied (os error 13)
   1 baddata
   1 digraph
   1 directive
   1 https
   2 remplace
   2 types
   3 there'snodirectivehere
   4 hazard
   4 note
   6 date
   8 class
   8 d
  12 dir
  16 r
  24 raw
 298 image
 448 replace
2656 unicode
```

I think more generally a substition def should be `.. |thename| directive-without-dot-dot:: content`, so we might be able to not special case and just recurse on directive ?

Reply:
That makes sense, I went with the docutils information but I guess there are other inline directives that are not docutils. I will give it a try. Marking this as a draft in the meantime